src/main/java/io/confluent/kafkarest/simpleconsumerspool/SizeLimitedSimpleConsumerPool.java

```diff
@@ -0,0 +1,91 @@
+/**
+ * Copyright 2015 Confluent Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ **/
+package io.confluent.kafkarest.simpleconsumerspool;
+
+import io.confluent.kafkarest.SimpleConsumerFactory;
+import io.confluent.kafkarest.SimpleFetcher;
+import kafka.javaapi.consumer.SimpleConsumer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.PriorityQueue;
+import java.util.Queue;
+
+/**
+ * The SizeLimitedSimpleConsumerPool keeps a pool of SimpleConsumers
+ * and can increase the pool within a specified limit
+ */
+public class SizeLimitedSimpleConsumerPool implements SimpleConsumerPool {
+  private static final Logger log = LoggerFactory.getLogger(SizeLimitedSimpleConsumerPool.class);
+
+  private final int maxPoolSize;
+  private final SimpleConsumerFactory simpleConsumerFactory;
+
+  private final Map<String, SimpleConsumer> simpleConsumers;
+  private final Queue<String> availableConsumers;
+
+  public SizeLimitedSimpleConsumerPool(int maxPoolSize, SimpleConsumerFactory simpleConsumerFactory) {
+    this.maxPoolSize = maxPoolSize;
+    this.simpleConsumerFactory = simpleConsumerFactory;
+
+    simpleConsumers = new HashMap<String, SimpleConsumer>();
+    availableConsumers = new PriorityQueue<String>();
+  }
+
+  synchronized public SimpleFetcher get(final String host, final int port) {
+
+    while (true) {
+      // If there is a SimpleConsumer available
+      if (availableConsumers.size() > 0) {
+        final String consumerId = availableConsumers.remove();
+        return new SimpleFetcher(simpleConsumers.get(consumerId), this);
+      }
+
+      // If not consumer is available, but we can instantiate a new one
+      if (simpleConsumers.size() < maxPoolSize) {
+        final SimpleConsumer simpleConsumer = simpleConsumerFactory.createConsumer(host, port);
+        simpleConsumers.put(simpleConsumer.clientId(), simpleConsumer);
+        return new SimpleFetcher(simpleConsumer, this);
+      }
+
+      // If no consumer is available and we reached the limit
+      try {
+        wait();
```

Comment:
It's possible this code could be simpler if you used a BlockingQueue with timeouts since they already support capacity restrictions. Might be worth trying out that version to see if it cuts down the code significantly.